File: 2023/day16/part2.ts
// Part 2 for day 16 of 2023

import { fillGrid } from "./part1";

export function part2(input: string): number {
  let grid = input
    .split("\n")
    .filter((c) => c != "")
    .map((c) => c.split(""));

  let total = 0;

  for (let i = 0; i < grid.length; i++) {
    total = Math.max(total, fillGrid(grid, [[[i, 0], "R"]]));
    total = Math.max(total, fillGrid(grid, [[[i, grid[0].length - 1], "L"]]));
  }

  for (let i = 0; i < grid[0].length; i++) {
    total = Math.max(total, fillGrid(grid, [[[0, i], "D"]]));
    total = Math.max(total, fillGrid(grid, [[[grid.length - 1, i], "U"]]));
  }

  return total;
}

File: 2024/src/day14/bin/part2.rs
//! # Advent of Code - Day 14 - Part Two

use std::collections::HashMap;

use crate::part1::{parse_robot, Robot};

pub fn part2(_input: &str, nr: i32, nc: i32) -> usize {
    let mut robots = _input
        .split("\n")
        .filter(|x| !x.is_empty())
        .map(|robot_str| parse_robot(robot_str))
        .collect::<Vec<Robot>>();

    let mut best_iter: i64 = -1;
    let mut min_sf = nr * nc * nr * nc;
    for i in 0..(nr * nc) {
        for robot in robots.iter_mut() {
            robot.move_robot(1, nr, nc);
        }
        let mut quadrants = HashMap::new();
        for robot in robots.iter_mut() {
            if robot.r < nr / 2 && robot.c < nc / 2 {
                *quadrants.entry(0).or_insert(0) += 1;
            } else if robot.r < nr / 2 && robot.c > nc / 2 {
                *quadrants.entry(1).or_insert(0) += 1;
            } else if robot.r > nr / 2 && robot.c > nc / 2 {
                *quadrants.entry(2).or_insert(0) += 1;
            } else if robot.r > nr / 2 && robot.c < nc / 2 {
                *quadrants.entry(3).or_insert(0) += 1;
            }
        }
        let mut total = 1;

        for (_, v) in quadrants.iter() {
            total *= v;
        }
        if total < min_sf {
            min_sf = total;
            best_iter = i as i64;
        }
    }

    return best_iter as usize + 1;
}

#[cfg(test)]
mod day14 {
    use super::*;

    #[test]
    fn test_part2() {
        let _input = include_str!("../testinput.txt");
        assert_eq!(part2(_input, 7, 11), 7);
    }
}
